Recognize non existent annotations as illegal

MDoerner opened the pull request:
This PR closes #3336.

I completely rewrote the `IllegalAnnotationsInspection` to work based on the resolves annotations and user declarations. This was necessary to get the correct scoping behaviour, which is not directly related to anything in the parse trees.

I also introduced a `NotRecognizedAnnotation` for all annotations not recognized by RD. Now, these are reported as illegal. Previously, they caused the inspection to throw and abort.

rkapka commented on Rubberduck.Parsing/Annotations/NotRecognizedAnnotation.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Rubberduck.Parsing.Annotations
+{
+    /// <summary>
```
Wrong comment.

rkapka commented on Rubberduck.Parsing/Annotations/NotRecognizedAnnotation.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Rubberduck.Parsing.Annotations
+{
+    /// <summary>
+    /// Used for specifying the Code Explorer folder a appears under.
```
Wrong comment.

MDoerner replied:
Yay, copy-pasta error.
Will fix

rkapka commented on Rubberduck.CodeAnalysis/Inspections/Concrete/IllegalAnnotationInspection.cs:
```diff
@@ -1,175 +1,64 @@
+        private ICollection<IAnnotation> NonMemberAnnotationsOnMembers(IEnumerable<Declaration> userDeclarations)
+        {
+            return userDeclarations
+                .Where(declaration => !declaration.DeclarationType.HasFlag(DeclarationType.Module) && declaration.DeclarationType != DeclarationType.Project)
```
Won't `.Where(declaration => declaration.DeclarationType.HasFlag(DeclarationType.Member)` do the same thing?

MDoerner replied:
No, it does not: e.g. variables and enumeration members do not have the `Member` flag.

rkapka replied:
Does that mean that a variable is not a member but can have a member annotation? If yes, that's misleading.

MDoerner commented on Rubberduck.Parsing/Annotations/AnnotationType.cs:
```diff
@@ -43,7 +43,12 @@ public enum AnnotationType
+        /// <summary>
+        /// A type for all not recognized annotations.
+        /// </summary>
+        NotRecognized = 1 << 19
```
Can do that

MDoerner commented:
I added more categories of annotation types and made the illegal annotations inspection more fine-grained using them.

This uncovered the problem that annotations on module level variables got added to the module as well. I changed that in a bit hacky way. Note that this may brake some malformed test modules like those in some tests in which the `@TestModule` annotation was right above the declaratio  of the `Assert` member. 

Annotations still have some problems like that identifier references in the declaration section are also applied to the module. However, I imagine that this is rather rare.

rkapka commented on Rubberduck.CodeAnalysis/Inspections/Concrete/IllegalAnnotationInspection.cs:
```diff
@@ -1,175 +1,112 @@
+            DeclarationType.UserDefinedTypeMember
+        };
+
+        private static ICollection<IAnnotation> NoneGeneralAnnotationsWhereNoOtherAnnotationsBelong(IEnumerable<Declaration> userDeclarations)
```
`None` -> `Non`

rkapka replied:
I don't this method is in line with the definition of `GeneralAnnotation`. Can you please explain what it does, because I don't understand.

MDoerner replied:
Arrgh, I forgot to add the last condition that the annotation does not have the general flag.

This also needs a test.

rkapka commented on Rubberduck.Parsing/Annotations/AnnotationType.cs:
```diff
@@ -43,7 +64,7 @@ public enum AnnotationType
         [AttributeAnnotation("VB_Exposed", "True")]
         Exposed = 1 << 17 | Attribute | ModuleAnnotation,
-        Obsolete = 1 << 18 | MemberAnnotation
+        Obsolete = 1 << 18 | MemberAnnotation | VariableAnnotation
```
Based on `VariableAnnotationDeclarationTypes` variables and constants can be marked as obsolete. Does it make sense?

rkapka replied:
I am talking about variables and constants appearing inside members.

MDoerner replied:
The declarations do not distinguish between variables and constants inside members and on module level. Since `Obsolete` makes sense for the later in my opinion I assembly the flag to the annotation type.

rkapka commented on RubberduckTests/Inspections/IllegalAnnotationsInspectionTests.cs:
```diff
@@ -580,6 +580,121 @@ End Sub
+
+        [Test]
+        [Category("Inspections")]
+        public void VariableAnnotationOnIdentifier_OneResult()
```
`Identifier` -> `Label`
